File: core/src/main/java/com/ataulm/Event.java
package com.ataulm;

import com.google.auto.value.AutoValue;

@AutoValue
public abstract class Event<T> {

    public static <T> Event<T> loading() {
        return loading(null);
    }

    public static <T> Event<T> loading(T data) {
        return new AutoValue_Event<>(Type.LOADING, Optional.fromNullable(data), Optional.<Throwable>absent());
    }

    public static <T> Event<T> idle() {
        return idle(null);
    }

    public static <T> Event<T> idle(T data) {
        return new AutoValue_Event<>(Type.IDLE, Optional.fromNullable(data), Optional.<Throwable>absent());
    }

    public static <T> Event<T> error(Throwable error) {
        return error(null, error);
    }

    public static <T> Event<T> error(T data, Throwable error) {
        if (error == null) {
            throw new IllegalArgumentException("Error events must contain an error.");
        }
        return new AutoValue_Event<>(Type.ERROR, Optional.fromNullable(data), Optional.of(error));
    }

    public abstract Type getType();

    public abstract Optional<T> getData();

    public abstract Optional<Throwable> getError();

    Event() {
        // instantiate AutoValue generated class
    }

    public enum Type {

        LOADING,
        ERROR,
        IDLE

    }

}

File: core/src/main/java/com/ataulm/Event.kt
package com.ataulm

data class Event<T>(val type: Type, val data: T?, val error: Throwable?) {

    fun updateData(data: T): Event<T> {
        return Event(type, data, error)
    }

    enum class Type {

        LOADING,
        ERROR,
        IDLE

    }

    companion object {

        fun <T> loading(): Event<T> {
            return loading(null)
        }

        fun <T> loading(data: T?): Event<T> {
            return Event(Type.LOADING, data, null)
        }

        fun <T> idle(): Event<T> {
            return idle(null)
        }

        fun <T> idle(data: T?): Event<T> {
            return Event(Type.IDLE, data, null)
        }

        fun <T> error(error: Throwable): Event<T> {
            return error(null, error)
        }

        fun <T> error(data: T?, error: Throwable?): Event<T> {
            if (error == null) {
                throw IllegalArgumentException("Error events must contain an error.")
            }
            return Event(Type.ERROR, data, error)
        }
    }

}
